common/src/IO/AssimpParser.cpp

```diff
@@ -327,6 +343,126 @@ std::optional<size_t> getMeshIndex(const aiScene& scene, const aiMesh& mesh)
+    {
+      return size_t(i);
+    }
+  }
+  return std::nullopt;
+}
+
+/**
+ * Computes the animation information for each channel in an
+ * animation sequence. Always uses the first frame of the animation.
+ */
+std::vector<AssimpBoneInformation> getAnimationInformation(
+  const aiNode& root, const aiAnimation& animation)
+{
+  auto indivTransforms = std::vector<aiMatrix4x4>{animation.mNumChannels};
+
+  // calculate the transformations for each animation channel
+  for (unsigned int i = 0; i < animation.mNumChannels; ++i)
+  {
+    const aiNodeAnim* channel = animation.mChannels[i];
+
+    aiVector3D position(0, 0, 0);
+    if (channel->mNumPositionKeys > 0)
+    {
+      position = channel->mPositionKeys[0].mValue;
+    }
+
+    aiQuaternion rotation(1, 0, 0, 0);
+    if (channel->mNumRotationKeys > 0)
+    {
+      rotation = channel->mRotationKeys[0].mValue;
+    }
+
+    aiVector3D scale(1, 1, 1);
+    if (channel->mNumScalingKeys > 0)
+    {
+      scale = channel->mScalingKeys[0].mValue;
+    }
+
+    // build a transformation matrix from it
+    aiMatrix4x4& mat = indivTransforms[i];
+    mat = aiMatrix4x4(rotation.GetMatrix());
+    mat.a1 *= scale.x;
+    mat.b1 *= scale.x;
+    mat.c1 *= scale.x;
+    mat.a2 *= scale.y;
+    mat.b2 *= scale.y;
+    mat.c2 *= scale.y;
+    mat.a3 *= scale.z;
+    mat.b3 *= scale.z;
+    mat.c3 *= scale.z;
+    mat.a4 = position.x;
+    mat.b4 = position.y;
+    mat.c4 = position.z;
+  }
+
+  // assemble the transform information from the bone hierarchy (child
+  // bones must be multiplied by their parent transformations, recursively)
+  auto transforms = std::vector<AssimpBoneInformation>{animation.mNumChannels};
+  for (unsigned int i = 0; i < animation.mNumChannels; ++i)
+  {
+    const aiNodeAnim* channel = animation.mChannels[i];
+
+    // start with the individual transformation of this channel
+    auto globalTransform = indivTransforms[i];
+    int32_t parentId = -1;
+
+    // traverse the bone hierarchy to compute the global transformation
+    auto boneNode = root.FindNode(channel->mNodeName);
+    if (!boneNode)
+    {
+      continue; // couldn't find the bone node, something is weird
+    }
+
+    // start at the first parent and walk up the tree
+    aiNode* parentNode = boneNode->mParent;
+    while (parentNode)
+    {
+      // use the node's default transformation, in case node isn't a bone
+      // and won't be transformed by this animation
+      aiMatrix4x4 nodeTransformation = parentNode->mTransformation;
+
+      // find the index of this bone in the channel list
+      const auto index = getChannelIndex(animation, *parentNode);
+      if (index.has_value())
+      {
+        nodeTransformation = indivTransforms[index.value()];
+
+        // if this is the first parent of the bone, set the parent id
+        if (parentNode == boneNode->mParent)
+        {
+          parentId = int32_t(index.value());
+        }
+      }
+      else
+      {
+        break;
```

Comment:
Could you extract function here so that we don't have this difficult to follow control flow? Maybe you could extract a function that does the search recursively instead of in a while loop.

Reply:
Done, although I'm not sure if my solution is the best one. I need to know the index of the immediate parent bone, as well as the full recursive transformation, so I returned tuple from the recursive helper to avoid traversing the tree twice.